### PR TITLE
Enhancement / expression function categoryItemProp

### PR DESCRIPTION
**Pull request by SteRiccio**
Changes in this pull request:
- added SurveyRefData type to Survey (it is used to store a sort of cache of category and taxon items)
- added categoryItemProp function to record expression evaluator
- function tested with unit tests
- added SurveyBuilder and RecordBuilder in tests to simplify test survey and record creation

### CODE REVIEW

File: src/survey/survey.ts [hunk 1]
```diff
@@ -44,4 +53,7 @@ export interface Survey extends ArenaObject<SurveyProps> {
   published: boolean
   readonly id?: number
   uuid: string
+  // indexes
+  categories: { [categoryUuid: string]: Category }
+  indexRefData: SurveyRefData
```
Comment (minotogna) [inline src/survey/survey.ts:58]: let's call this just refDdata ?

Comment (minotogna) [reply src/survey/survey.ts:58]: it's optional

File: src/survey/factory.ts [hunk 2]
```diff
@@ -52,6 +52,13 @@ export const SurveyFactory: Factory<Survey, SurveyFactoryParams> = {
         descriptions,
         collectUri,
       },
+      categories: {},
+      indexRefData: {
+        categoryItemUuidIndex: {},
+        categoryItemIndex: {},
+        taxonIndex: {},
+        taxonUuidIndex: {},
+      },
```
Comment (minotogna) [inline src/survey/factory.ts:61]: you don't need this

File: src/tests/builder/surveyBuilder/surveyBuilder.ts [hunk 3]
```diff
@@ -1,31 +1,52 @@
...
+      items.forEach((item) => {
+        Objects.setInPath({
+          obj: survey.indexRefData.categoryItemUuidIndex,
+          value: item.uuid,
+          path: [category.uuid, item.parentUuid || 'null', item.props.code || ''],
+        })
+        survey.indexRefData.categoryItemIndex[item.uuid] = item
+      })
+      survey.categories[category.uuid] = category
+    })
```
Comment (minotogna) [inline src/tests/builder/surveyBuilder/surveyBuilder.ts:49]: we need a SurveyRefDataFactory

**Review by minotogna — CHANGES_REQUESTED** [review]

File: src/survey/survey.ts [hunk 4]
```diff
@@ -33,6 +35,13 @@ export interface SurveyProps {
   collectUri?: string
 }
 
+export interface SurveyRefData {
```
Comment (minotogna) [inline src/survey/survey.ts:38]: it must be exported from survey/index

File: src/survey/survey.ts (hunk 1 above)
Comment (SteRiccio) [reply src/survey/survey.ts:58]: ok

File: src/survey/factory.ts (hunk 2 above)
Comment (SteRiccio) [reply src/survey/factory.ts:61]: ok

File: src/tests/builder/surveyBuilder/surveyBuilder.ts (hunk 3 above)
Comment (SteRiccio) [reply src/tests/builder/surveyBuilder/surveyBuilder.ts:49]: right

File: src/survey/survey.ts (hunk 4 above)
Comment (SteRiccio) [reply src/survey/survey.ts:38]: ok

**Review by minotogna — DISMISSED** [review]

**Review by minotogna — APPROVED** [review]

**Review by minotogna — APPROVED** [review]

**Review by sorja — APPROVED** [review]

**Pull request merged**